# 宏任务和微任务：不是所有任务都是一个待遇
## 宏任务
在消息队列中的任务称为**宏任务**，消息队列中的任务是通过事件循环系统来执行的。  

**消息队列中宏任务的执行过程**
- 先从多个消息队列中选出一个最老的任务，这个任务称为oldestTask
- 然后循环系统记录任务开始执行的时间，并把这个oldestTask设置为当前正在执行的任务
- 当任务执行完成后，删除当前正在执行的任务，并从对应的消息队列中删除掉这个oldestTask
- 最后统计执行完成的时长等消息

```
<!DOCTYPE html>
<html>
    <body>
        <div id='demo'>
            <ol>
                <li>test</li>
            </ol>
        </div>
    </body>
    <script type="text/javascript">
        function timerCallback2(){
          console.log(2)
        }
        function timerCallback(){
            console.log(1)
            setTimeout(timerCallback2,0)
        }
        setTimeout(timerCallback,0)
    </script>
</html>

```
![](img/宏任务.png)  

**setTimeout函数触发的回调函数都是宏任务**

## 微任务
**微任务就是一个需要异步执行的函数，执行时机是在主函数执行结束之后、当前宏任务结束之前**  

JS执行一段脚本时，V8会为其创建一个全局执行上下文，在创建执行上下文的同时，V8引擎也会在内部创建一个**微任务队列**。用来存放当前宏任务执行过程中产生的多个微任务。  

### 产生微任务的两种方式：  
1. 使用**MutationObserver监控**某个DOM节点，然后再通过JavaScript来修改这个节点，或者为这个节点添加、删除部分子节点，当DOM节点发生变化时，就会产生DOM变化记录的微任务。
2. **使用Promise**,当调用Promise.resolve()或Promise.reject()时，也会产生微任务。


微任务会被JS引擎按照顺序保存到微任务队列中。


### 微任务队列何时被执行
在当前宏任务中的JavaScript快执行完成时，也就是JS引擎准备退出全局执行上下文并清空调用栈的时候，JS引擎会检查全局执行上下文中的微任务队列，然后按照顺序执行队列中的微任务。  

**WHATWG把执行微任务的时间点称为检查点**  

在执行微任务过程中产生的新的微任务不会推迟到下个宏任务中执行，而是在当前的宏任务中继续执行。  

![](img/添加微任务.png)  

![](img/执行微任务.png)  

- 微任务和宏任务是绑定的，每个宏任务在执行时，会创建自己的微任务队列
- 微任务的执行时长会影响到当前宏任务的时长。比如一个宏任务在执行过程中，产生了100个微任务，执行每个微任务的时间是10毫秒，那么执行这100个微任务的时间就是1000毫秒，这100个微任务就让宏任务的执行时间延长了1000毫秒。
- 在一个宏任务中，分别创建一个用于回调的宏任务和微任务，无论什么情况下，微任务都早于宏任务执行

## 监听DOM变化方法
利用MutationObserver API来监听DOM的变化，包括属性的变化、节点的增减、内容的变化等。  
1. MutationObserver将响应函数改成异步调用，不在每次DOM变化时都触发异步调用，而是在多次变化后，**一次触发异步调用**，并使用一个数据结构来记录这期间所有的DOM变化。
2. 在每次DOM节点发生变化的时候，渲染引擎将变化记录封装成微任务，并将微任务添加进当前的微任务队列中。这样当执行到检查点时，V8引擎就会按照顺序执行微任务了。


MutationObserver采用了“**异步+微任务**”的策略  
- 通过异步操作解决了同步操作的性能问题
- 通过微任务解决了实时性的问题


## 思考题
```
function executor(resolve, reject) {
    let rand = Math.random();
    console.log(1)
    console.log(rand)
    if (rand > 0.5)
        resolve()
    else
        reject()
}
var p0 = new Promise(executor);

var p1 = p0.then((value) => {
    console.log("succeed-1")
    return new Promise(executor)
})


var p3 = p1.then((value) => {
    console.log("succeed-2")
    return new Promise(executor)
})

var p4 = p3.then((value) => {
    console.log("succeed-3")
    return new Promise(executor)
})


p4.catch((error) => {
    console.log("error")
})
console.log(2)

```

四个Promise对象：p0~p4,无论哪个对象里面抛出异常，都可以通过最后一个对象p4.catch来捕获异常。  
**因为Promise对象的错误具有“冒泡”性质，会一直向后传递直到被onReject函数处理或catch语句捕获为止。**